bank2/main.go

```diff
@@ -0,0 +1,267 @@
+				return err
+			}
+			update := `UPDATE account SET balance = $1 WHERE id = $2;`
+			if _, err = tx.Exec(update, toBalance+amount, to); err != nil {
+				return err
+			}
+			if _, err = tx.Exec(update, fromBalance-amount, from); err != nil {
+				return err
+			}
+			writeDuration = time.Since(startWrite)
+			return nil
+		}); err != nil {
+			log.Printf("failed transaction: %v", err)
+			continue
+		}
+		if fromBalance >= amount {
+			readings <- measurement{read: readDuration, write: writeDuration, total: time.Since(start)}
+		}
+	}
+}
+
+func verifyTotalBalance(db *sql.DB) {
+	var sum int
+	if err := db.QueryRow("SELECT SUM(balance) FROM account").Scan(&sum); err != nil {
+		log.Fatal(err)
+	}
+	if sum != *numAccounts*initialBalance+initialSystemBalance {
+		log.Printf("The total balance is incorrect: %d.", sum)
+		os.Exit(1)
+	}
+}
+
+var usage = func() {
+	fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
+	fmt.Fprintf(os.Stderr, "  %s <db URL>\n\n", os.Args[0])
+	flag.PrintDefaults()
+}
+
+func main() {
+	flag.Usage = usage
+	flag.Parse()
+
+	dbURL := "postgresql://root@localhost:26257/bank2?sslmode=disable"
+	if flag.NArg() == 1 {
+		dbURL = flag.Arg(0)
+	}
+
+	parsedURL, err := url.Parse(dbURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	parsedURL.Path = "bank2"
+
+	db, err := sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() { _ = db.Close() }()
+
+	if _, err := db.Exec("CREATE DATABASE IF NOT EXISTS bank2"); err != nil {
+		log.Fatal(err)
+	}
+
+	// concurrency + 1, for this thread and the "concurrency" number of
+	// goroutines that move money
+	db.SetMaxOpenConns(*concurrency + 1)
+
+	if _, err = db.Exec(`
+CREATE TABLE IF NOT EXISTS account (
+  id INT,
+  balance BIGINT NOT NULL,
+  name STRING,
+
+  PRIMARY KEY (id),
+  UNIQUE INDEX byName (name)
+);
+
+CREATE TABLE IF NOT EXISTS transaction (
+  id INT,
+  booking_date TIMESTAMP DEFAULT NOW(),
+  txn_date TIMESTAMP DEFAULT NOW(),
+  txn_ref STRING,
+
+  PRIMARY KEY (id),
+  UNIQUE INDEX byTxnRef (txn_ref)
+);
+
+CREATE TABLE IF NOT EXISTS transaction_leg (
+  id BYTES DEFAULT uuid_v4(),
+  account_id INT,
+  amount BIGINT NOT NULL,
+  running_balance BIGINT NOT NULL,
+  txn_id INT,
+
+  PRIMARY KEY (id)
+);
+
+TRUNCATE TABLE account;
+TRUNCATE TABLE transaction;
+TRUNCATE TABLE transaction_leg;
+`); err != nil {
+		log.Fatal(err)
+	}
+
+	insertSQL := "INSERT INTO account (id, balance, name) VALUES ($1, $2, $3)"
+
+	// Insert initialSystemBalance into the system account.
+	if _, err = db.Exec(insertSQL, systemAccountID, initialSystemBalance, "system account"); err != nil {
+		log.Fatal(err)
+	}
+	// Insert initialBalance into all user accounts.
+	for i := 1; i <= *numAccounts; i++ {
+		if _, err = db.Exec(insertSQL, i, initialBalance, fmt.Sprintf("account %d", i)); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	verifyTotalBalance(db)
+
+	lastNow := time.Now()
+	readings := make(chan measurement, 10000)
```

Comment:
I realize this is copied from the old `bank` example. Looking at it now, I think this would be better as a single struct that is passed to `moveMoney` and for the fields to be incremented atomically.